spec/unit/provider/batch_spec.rb

```diff
@@ -0,0 +1,130 @@
+    before do
+      allow(ChefUtils).to receive(:windows?).and_return(true)
+      stub_const("Chef::ReservedNames::Win32::API::Security::GENERIC_READ", 1)
+      stub_const("Chef::ReservedNames::Win32::API::Security::GENERIC_EXECUTE", 4)
+      stub_const("Chef::ReservedNames::Win32::Security", Class.new)
+      stub_const("Chef::ReservedNames::Win32::Security::SecurableObject", Class.new)
+      stub_const("Chef::ReservedNames::Win32::Security::SID", Class.new)
+      stub_const("Chef::ReservedNames::Win32::Security::ACE", Class.new)
+      stub_const("Chef::ReservedNames::Win32::Security::ACL", Class.new)
+
+      provider.singleton_class.send(:public, :grant_alternate_user_read_access)
+    end
+
+    context "when an alternate user is not specified" do
+      it "does not attempt to set the script file's security descriptor" do
+        expect(provider).to receive(:grant_alternate_user_read_access)
+        expect(Chef::ReservedNames::Win32::Security::SecurableObject).not_to receive(:new)
+        provider.grant_alternate_user_read_access("a fake path")
+      end
+    end
+
+    context "when an alternate user is specified" do
+      let(:security_descriptor) { instance_double("Chef::ReservedNames::Win32::Security::SecurityDescriptor", dacl: []) }
+      let(:securable_object) { instance_double("Chef::ReservedNames::Win32::Security::SecurableObject", :security_descriptor => security_descriptor, :dacl= => nil) }
+
+      it "sets the script file's security descriptor" do
+        new_resource.user("toor")
+        expect(Chef::ReservedNames::Win32::Security::SecurableObject).to receive(:new).and_return(securable_object)
+        expect(Chef::ReservedNames::Win32::Security::SID).to receive(:from_account).and_return(nil)
+        expect(Chef::ReservedNames::Win32::Security::ACE).to receive(:access_allowed).and_return(nil)
+        expect(Chef::ReservedNames::Win32::Security::ACL).to receive(:create).and_return(nil)
+        expect(securable_object).to receive(:dacl=)
+        provider.grant_alternate_user_read_access("a fake path")
+      end
+    end
+  end
+
+  describe "#with_temp_script_file" do
+    before do
+      provider.singleton_class.send(:public, :with_temp_script_file)
+      provider.singleton_class.send(:public, :script_file_path)
+    end
+
+    it "should put the contents of the script in the temp file" do
+      temp_file_contents = nil
+
+      provider.with_temp_script_file do
+        temp_file_contents = File.read(provider.script_file_path)
+      end
+
+      expect(temp_file_contents.strip).to eq(%q{echo "hello"})
+    end
+  end
+
+  describe "#command" do
+    let(:basepath) { "C:\\Windows\\system32" }
+    let(:interpreter) { File.join(basepath, "cmd.exe") }
+
+    before do
+      allow(provider).to receive(:basepath).and_return(basepath)
+      provider.singleton_class.send(:public, :with_temp_script_file)
+      provider.singleton_class.send(:public, :script_file_path)
```

Comment:
This is kinda gross but I thought it was better than a bunch of stubbing and mocking.